commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/CommitStatusPublisherListener.java

```diff
@@ -172,47 +177,85 @@ public boolean run(@NotNull CommitStatusPublisher publisher, @NotNull BuildRevis
+      runForEveryPublisher(event, bp.getBuildType(), bp.getAssociatedBuild(), task, description, publishedRevisions);
+    }
+
+  }
+
+  private void runForEveryPublisher(@NotNull Event event, @NotNull SBuildType buildType, @NotNull SBuild build, @NotNull PublishTask task, @NotNull String description, @NotNull Set<BuildRevision> publishedRevisions) {
     if (build.isPersonal()) {
-      for(SVcsModification change: build.getBuildPromotion().getPersonalChanges()) {
+      for (SVcsModification change : build.getBuildPromotion().getPersonalChanges()) {
         if (change.isPersonal())
           return;
       }
     }
     Map<String, CommitStatusPublisher> publishers = getPublishers(buildType);
-    LOG.debug("Event: " + event.getName() + ", build " + LogUtil.describe(build) + ", publishers: " + publishers.values());
+    LOG.debug("Event: " + event.getName() + ", build " + LogUtil.describe(build) + ", publishers: " + publishers.values() + ", description: " + description);
     for (Map.Entry<String, CommitStatusPublisher> pubEntry : publishers.entrySet()) {
       CommitStatusPublisher publisher = pubEntry.getValue();
       if (!publisher.isEventSupported(event))
         continue;
       if (isPublishingDisabled(buildType)) {
-        logStatusNotPublished(event, LogUtil.describe(build), publisher, "commit status publishing is disabled");
+        logStatusNotPublished(event, description, publisher, "commit status publishing is disabled");
         continue;
       }
       List<BuildRevision> revisions = getBuildRevisionForVote(publisher, build);
       if (revisions.isEmpty()) {
-        logStatusNotPublished(event, LogUtil.describe(build), publisher, "no compatible revisions found");
+        logStatusNotPublished(event, description, publisher, "no compatible revisions found");
         continue;
       }
       myProblems.clearProblem(publisher);
-      for (BuildRevision revision: revisions) {
-        runTask(event, build.getBuildPromotion(), LogUtil.describe(build), task, publisher, revision);
+      for (BuildRevision revision : revisions) {
+        if (!publishedRevisions.contains(revision)) {
+          runTask(event, build.getBuildPromotion(), description, task, publisher, revision);
+          publishedRevisions.add(revision);
+        }
+
       }
     }
     myProblems.clearObsoleteProblems(buildType, publishers.keySet());
   }
 
   private void runForEveryPublisherQueued(@NotNull Event event, @NotNull SBuildType buildType, @NotNull SQueuedBuild build, @NotNull PublishTask task) {
+    final String description = LogUtil.describe(build);
+    Set<BuildRevision> publishedRevisions = new HashSet<BuildRevision>();
+    runForEveryPublisherQueued(event, buildType, build, task, description, publishedRevisions);
+
+    if (!isPublishingToDependenciesEnabled(buildType)) {
+      return;
+    }
+
+    for (BuildPromotion bp : build.getBuildPromotion().getAllDependencies()) {
```

Comment:
It seems to me it doesn't make sense to propagate added to/removed from queue events to revisions from dependencies. There is also seemingly an uncertainty wether all dependencies actually have queued builds, or some have already running/finished ones.